src/rocksdb.rs

```diff
@@ -39,16 +40,134 @@ pub struct ReadOptions {
+            let val = unsafe { slice::from_raw_parts(val_ptr, val_len as usize) };
+            Some((key,val))
+        } else {
+            None
+        }
+    }
+}
+
+impl DBIterator {
+//TODO alias db & opts to different lifetimes, and DBIterator to the db's lifetime 
+    fn new(db: &RocksDB, readopts: &ReadOptions) -> DBIterator {
+        unsafe {
+            let iterator = rocksdb_ffi::rocksdb_create_iterator(db.inner, readopts.inner);
+            rocksdb_ffi::rocksdb_iter_seek_to_first(iterator);
+            DBIterator{ inner: iterator, direction: Direction::forward, just_seeked: true }
+        }
+    }
+
+    pub fn from_start(&mut self) -> SubDBIterator {
+        self.just_seeked = true;
+        unsafe {
+            rocksdb_ffi::rocksdb_iter_seek_to_first(self.inner);
+        };
+        SubDBIterator{ iter: self, direction: Direction::forward, }
+    }
+
+    pub fn from_end(&mut self) -> SubDBIterator {
+        self.just_seeked = true;
+        unsafe {
+            rocksdb_ffi::rocksdb_iter_seek_to_last(self.inner);
+        };
+        SubDBIterator{ iter: self, direction: Direction::reverse, }
+    }
+
+    pub fn from(&mut self, key: &[u8], dir: Direction) -> SubDBIterator {
+        self.just_seeked = true;
+        unsafe {
+            rocksdb_ffi::rocksdb_iter_seek(self.inner, key.as_ptr(), key.len() as size_t);
+        }
+        SubDBIterator{ iter: self, direction: dir, }
+    }
+}
+
+impl Drop for DBIterator {
+    fn drop(&mut self) {
+        unsafe {
+            rocksdb_ffi::rocksdb_iter_destroy(self.inner);
+        }
+    }
+}
+
+impl <'a> Snapshot<'a> {
+    pub fn new(db: &RocksDB) -> Snapshot {
+        let snapshot = unsafe { rocksdb_ffi::rocksdb_create_snapshot(db.inner) };
+        Snapshot{db: db, inner: snapshot}
+    }
+
+    pub fn iterator(&self) -> DBIterator {
+        let mut readopts = ReadOptions::new();
+        readopts.set_snapshot(self);
+        DBIterator::new(self.db, &readopts)
+    }
+}
+
+impl <'a> Drop for Snapshot<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            rocksdb_ffi::rocksdb_release_snapshot(self.db.inner, self.inner);
+        }
+    }
+}
+
 // This is for the RocksDB and write batches to share the same API
 pub trait Writable {
-    fn put(&self, key: &[u8], value: &[u8]) -> Result<(), String>;
-    fn merge(&self, key: &[u8], value: &[u8]) -> Result<(), String>;
-    fn delete(&self, key: &[u8]) -> Result<(), String>;
+    fn put(&mut self, key: &[u8], value: &[u8]) -> Result<(), String>;
+    fn merge(&mut self, key: &[u8], value: &[u8]) -> Result<(), String>;
+    fn delete(&mut self, key: &[u8]) -> Result<(), String>;
 }
 
 fn error_message(ptr: *const i8) -> String {
     let c_str = unsafe { CStr::from_ptr(ptr) };
-    from_utf8(c_str.to_bytes()).unwrap().to_owned()
+    let s = from_utf8(c_str.to_bytes()).unwrap().to_owned();
+    unsafe{
+        libc::free(ptr as *mut libc::c_void);
```

Comment:
I'm not sure if you're using OSX Yosemite or not, but in case you are you can install valgrind with `brew install --HEAD valgrind` (without --HEAD it does not work for this OS).